### ISSUE TITLE
Spurious "Extracting tar content of undefined failed"

### ISSUE THREAD

**Issue by jleclanche** [yarnpkg/yarn#7212]
Followup on #6312. Potentially related issues: #7087, #6755, #6312

#### Details
- `yarn --version`: 1.15.2
- [yarn.lock](https://github.com/yarnpkg/yarn/files/3100415/yarn.lock.txt)
- command-line: `rm ~/.cache/yarn ~/.npm ./node_modules -rf && yarn install --production --frozen-lockfile`
- Arch Linux. Also reproduced on `node:latest` docker image.

----

I can still reproduce the issue on the latest yarn version. I *do* have a git dependency (`"@magicfind/wf-build-engine": "https://gitlab.com/magicfind/warframe/wf-build-engine.git"`) and I'm fairly certain it's related to that, because I've been messing with it, though I don't know what exactly caused this to start happening.
I've been hitting #5235 and trying to work around it, and while doing that started hitting #6312.

Locally I see three different behaviours:
- I hit the error and yarn aborts
- I hit the error and yarn continues
- I do not hit the error and the install succeeds.

This points to a race condition of some type?

#### Example outputs

All outputs below run the exact same command line, removing all the cache I can find. I've not found a reliable way to reproduce the issue 100% of the time.

```
yarn install v1.15.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[1/4] Resolving packages...
[2/4] Fetching packages...
error https://registry.yarnpkg.com/require-main-filename/-/require-main-filename-1.0.1.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, chmod '/home/adys/.cache/yarn/v4/npm-require-main-filename-1.0.1-97f717b69d48784f5f526a6c5aa8ffdda055a4d1/node_modules/require-main-filename/LICENSE.txt'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

```
yarn install v1.15.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.4: The platform "linux" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
$ npm run build
npm WARN lifecycle The node binary used for scripts is /tmp/yarn--1555780098541-0.2687961721890739/node but npm is using /usr/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

> wf-build-engine@0.1.0 build /home/adys/.cache/yarn/v4/.tmp/83dda1d0ded23f5a9d7b82676a04465a.c67a312e2664e1fa7b0ae1176d36d03b6e40e079.prepare
> rollup -c


src/index.ts → dist/index.js, dist/index.es.js...                                                                                                                                             
created dist/index.js, dist/index.es.js in 1.1s
info fsevents@1.2.7: The platform "linux" is incompatible with this module.
info "fsevents@1.2.7" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.8: The platform "linux" is incompatible with this module.
info "fsevents@1.2.8" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > react-ga@2.5.7" has unmet peer dependency "prop-types@^15.6.0".
warning " > react-mde@7.3.2" has unmet peer dependency "draft-js@^0.10.5".
warning "react-scripts > eslint-config-react-app@4.0.0-next.b0cbf2ca" has incorrect peer dependency "babel-eslint@9.x".
warning "react-scripts > eslint-config-react-app@4.0.0-next.b0cbf2ca" has incorrect peer dependency "eslint-plugin-react-hooks@1.5.0".
warning " > react-table@6.9.2" has unmet peer dependency "prop-types@^15.5.0".
warning " > remark-shortcodes@0.2.1" has unmet peer dependency "remark-parse@>=3.0.0".
warning " > webpackbar@3.2.0" has unmet peer dependency "webpack@^3.0.0 || ^4.0.0".
[4/4] Building fresh packages...
Done in 23.64s.
```

```
yarn install v1.15.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[1/4] Resolving packages...
[2/4] Fetching packages...
error https://registry.yarnpkg.com/is-stream/-/is-stream-1.1.0.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, open '/home/adys/.cache/yarn/v4/npm-is-stream-1.1.0-12d4a3dd4e68e0b79ceb8dbc84173ae80d91ca44/node_modules/is-stream/package.json'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

```
yarn install v1.15.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[1/4] Resolving packages...
[2/4] Fetching packages...
error https://registry.yarnpkg.com/iconv-lite/-/iconv-lite-0.4.24.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, chmod '/home/adys/.cache/yarn/v4/npm-iconv-lite-0.4.24-2022b4b25fbddc21d2f524974a474aafe733908b/node_modules/iconv-lite/encodings/utf16.js'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
info fsevents@1.2.4: The platform "linux" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
$ npm run build
npm WARN lifecycle The node binary used for scripts is /tmp/yarn--1555780287695-0.048757454583545856/node but npm is using /usr/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

> wf-build-engine@0.1.0 build /home/adys/.cache/yarn/v4/.tmp/83dda1d0ded23f5a9d7b82676a04465a.c67a312e2664e1fa7b0ae1176d36d03b6e40e079.prepare
> rollup -c


src/index.ts → dist/index.js, dist/index.es.js...                                                                                                                                             
created dist/index.js, dist/index.es.js in 1.1s
```

**Comment by jleclanche** [comment]
Weird:

```
$ yarn cache list

yarn cache v1.15.2
error An unexpected error occurred: "There should only be one folder in a package cache (got  in /home/adys/.cache/yarn/v4/npm-@magicfind-wf-build-engine-0.1.0-c67a312e2664e1fa7b0ae1176d36d03b6e40e079/node_modules/@magicfind)".
info If you think this is a bug, please open a bug report with the information provided in "/home/adys/src/overframe/frontend/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/cache for documentation about this command.
```

```
tree ~/.cache/yarn/v4/npm-@magicfind-wf-build-engine-0.1.0-c67a312e2664e1fa7b0ae1176d36d03b6e40e079/
/home/adys/.cache/yarn/v4/npm-@magicfind-wf-build-engine-0.1.0-c67a312e2664e1fa7b0ae1176d36d03b6e40e079/
└── node_modules
    └── @magicfind

2 directories, 0 files
```

**Comment by jleclanche** [comment]
[yarn-error.log](https://github.com/yarnpkg/yarn/files/3100435/yarn-error.log)

**Comment by jleclanche** [comment]
The `undefined` showing up in the error message made me wonder if there was another issue at play here but it seems to be a missing parameter which is only used for the error message.

https://github.com/yarnpkg/yarn/blob/4ad88165fdf607a70c75db556a3af95864b2189a/src/fetchers/tarball-fetcher.js#L246

Padding `tarballCachePath` as a third parameter here will make the error message make more sense:

```
error https://registry.yarnpkg.com/glob/-/glob-7.1.3.tgz: Extracting tar content of "/home/adys/.cache/yarn/v4/npm-glob-7.1.3-3960832d3f1574108342dafd3a67b332c0969df1/node_modules/glob/.yarn-tarball.tgz" failed, the file appears to be corrupt: "ENOENT: no such file or directory, open '/home/adys/.cache/yarn/v4/npm-glob-7.1.3-3960832d3f1574108342dafd3a67b332c0969df1/node_modules/glob/changelog.md'"
```

It does nothing to actually help the error though.

**Comment by jleclanche** [comment]
I ran `rm ~/.cache/yarn ~/.npm node_modules yarn.lock && yarn install --verbose`, to see if anything else can be gleaned from there:

```
verbose 13.540595951 Error: https://registry.yarnpkg.com/fast-json-stable-stringify/-/fast-json-stable-stringify-2.0.0.tgz: Extracting tar content of "/home/adys/.cache/yarn/v4/npm-fast-json-stable-stringify-2.0.0-d5142c0caee6b1189f87d3a76111064f86c8bbf2/node_modules/fast-json-stable-stringify/.yarn-tarball.tgz" failed, the file appears to be corrupt: "ENOENT: no such file or directory, stat '/home/adys/.cache/yarn/v4/npm-fast-json-stable-stringify-2.0.0-d5142c0caee6b1189f87d3a76111064f86c8bbf2/node_modules/fast-json-stable-stringify/test/to-json.js'"
    at MessageError.ExtendableBuiltin (/usr/lib/node_modules/yarn/lib/cli.js:721:66)
    at new MessageError (/usr/lib/node_modules/yarn/lib/cli.js:750:123)
    at Extract.<anonymous> (/usr/lib/node_modules/yarn/lib/cli.js:62682:14)
    at Extract.emit (events.js:198:15)
    at Extract.module.exports.Extract.destroy (/usr/lib/node_modules/yarn/lib/cli.js:139624:17)
    at onunlock (/usr/lib/node_modules/yarn/lib/cli.js:139501:26)
    at /usr/lib/node_modules/yarn/lib/cli.js:45391:25
    at /usr/lib/node_modules/yarn/lib/cli.js:45357:23
    at /usr/lib/node_modules/yarn/lib/cli.js:62632:13
    at FSReqCallback.oncomplete (fs.js:158:21)
error https://registry.yarnpkg.com/fast-json-stable-stringify/-/fast-json-stable-stringify-2.0.0.tgz: Extracting tar content of "/home/adys/.cache/yarn/v4/npm-fast-json-stable-stringify-2.0.0-d5142c0caee6b1189f87d3a76111064f86c8bbf2/node_modules/fast-json-stable-stringify/.yarn-tarball.tgz" failed, the file appears to be corrupt: "ENOENT: no such file or directory, stat '/home/adys/.cache/yarn/v4/npm-fast-json-stable-stringify-2.0.0-d5142c0caee6b1189f87d3a76111064f86c8bbf2/node_modules/fast-json-stable-stringify/test/to-json.js'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

Full log: [verbose.txt.gz](https://github.com/yarnpkg/yarn/files/3100649/verbose.txt.gz)

**Comment by melaku-z** [comment]
related to #6312 
- Upgrade to 1.13+

**Comment by jleclanche** [comment]
This is on 1.15.2.

**Comment by melaku-z** [comment]
> This is on 1.15.2.

Yes, my bad. "yarn install --network-concurrency 1" is working for me, a temporary, slow workaround.

**Comment by kelvin1984** [comment]
I used the 1.7 version before today, it never happened, and today I upgraded to the latest version (1.15.2), it will happen: Extracting tar content of undefined failed, the file appears to be corrupt

**Comment by TikiTDO** [comment]
I was seeing this fairly consistently on 1.16.0, and was finally able to track this down.

Our setup is as follows:

1. A `/package.json` which defines a bunch of workspaces
2. A `/workspace/package.json` which references a git modules
3. A custom build of a node module in git, with a `package.json` with a `scripts -> prepare` key.

This was causing an extra instance of yarn to run that script, which then tried to fetch packages needed by this lib outside of the expected order. Since the lib in question is mostly dead, and barely needed in our case I was able to resolve the problem by removing the prepare script. 

For everyone else having these problems I recommend removing any git packages you have referenced, and failing that to grep all the package.json files in your project for [these reserved scripts](https://yarnpkg.com/en/docs/package-json#toc-scripts) since they may be the cause.

Hope this helps root cause this in a more reliable fashion.

**Comment by TikiTDO** [comment]
@arcanis You mentioned in #6312 that it was closed because there was no further details. I believe my post above this one may explain what's happening here. It look like (as an educated guess) that the instance of yarn being called to service the special scripts like `install`, `postinstall`, `prepublish`, and `prepare` is not respecting any sort of mutex config.

**Comment by foodforarabbit** [comment]
> I was seeing this fairly consistently on 1.16.0, and was finally able to track this down.
> 
> Our setup is as follows:
> 
> 1. A `/package.json` which defines a bunch of workspaces
> 2. A `/workspace/package.json` which references a git modules
> 3. **A custom build of a node module in git, with a `package.json` with a `scripts -> prepare` key.**
> 
> This was causing an extra instance of yarn to run that script, which then tried to fetch packages needed by this lib outside of the expected order. Since the lib in question is mostly dead, and barely needed in our case I was able to resolve the problem by removing the prepare script.
> 
> For everyone else having these problems I recommend removing any git packages you have referenced, and failing that to grep all the package.json files in your project for [these reserved scripts](https://yarnpkg.com/en/docs/package-json#toc-scripts) since they may be the cause. This is almost certainly the case if you see this sort of duplication:
> 
> ```
> [1/4] Resolving packages...
> [2/4] Fetching packages...
> [1/4] Resolving packages...
> [2/4] Fetching packages...
> ```
> 
> Hope this helps root cause this in a more reliable fashion.


THANK YOU! I had recently added .yarnrc with `--install.mutex network` to fix another issue (multiple simultaneous yarn in different repo on the same build server) and having this similar-but-shouldn't-be-possible-anymore error popped up really freaked me out. I noticed the repeated [1/4]... stuff though and turns out packages recursively calling yarn was indeed the problem!

I started by piping yarn into a file e.g. `yarn --verbose 2>&1 | tee yarn-output` and looking for clues around the area where I saw [1/4] repeat itself:

```
verbose 34.141 Performing "GET" request to "https://registry.yarnpkg.com/synchronous-promise/-/synchronous-promise-2.0.7.tgz".
verbose 34.15 Performing "GET" request to "https://registry.yarnpkg.com/toposort/-/toposort-2.0.2.tgz".
verbose 34.592 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/e237f3006359452a9561519898e180fb.5025ebf2427d709adfb10ee07f7dc36517d03694.prepare/.npmrc".
verbose 34.592 Checking for configuration file "/Users/xxxxx/.npmrc".
verbose 34.592 Checking for configuration file "/Users/xxxxx/.nvm/versions/node/v10.15.0/etc/npmrc".
verbose 34.592 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/e237f3006359452a9561519898e180fb.5025ebf2427d709adfb10ee07f7dc36517d03694.prepare/.npmrc".
verbose 34.592 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/.npmrc".
verbose 34.592 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.npmrc".
verbose 34.593 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/.npmrc".
verbose 34.593 Checking for configuration file "/Users/xxxxx/Library/Caches/.npmrc".
verbose 34.593 Checking for configuration file "/Users/xxxxx/Library/.npmrc".
verbose 34.593 Checking for configuration file "/Users/xxxxx/.npmrc".
verbose 34.593 Checking for configuration file "/Users/.npmrc".
verbose 34.593 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/e237f3006359452a9561519898e180fb.5025ebf2427d709adfb10ee07f7dc36517d03694.prepare/.yarnrc".
verbose 34.593 Checking for configuration file "/Users/xxxxx/.yarnrc".
verbose 34.593 Found configuration file "/Users/xxxxx/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/.nvm/versions/node/v10.15.0/etc/yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/e237f3006359452a9561519898e180fb.5025ebf2427d709adfb10ee07f7dc36517d03694.prepare/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/v4/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/Library/Caches/Yarn/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/Library/Caches/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/Library/.yarnrc".
verbose 34.594 Checking for configuration file "/Users/xxxxx/.yarnrc".
verbose 34.595 Found configuration file "/Users/xxxxx/.yarnrc".
verbose 34.595 Checking for configuration file "/Users/.yarnrc".
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/4] Resolving packages...
[2/4] Fetching packages...
verbose 34.869 Performing "GET" request to "https://registry.yarnpkg.com/core-js/-/core-js-2.5.7.tgz".
verbose 34.87 Performing "GET" request to "https://registry.yarnpkg.com/css-box-model/-/css-box-model-1.0.0.tgz".
```

luckily, I decided to check out **/Users/xxxxx/Library/Caches/Yarn/v4/.tmp/e237f3006359452a9561519898e180fb**.5025ebf2427d709adfb10ee07f7dc36517d03694.prepare/.yarnrc and that folder was still around - checked out the package.json and realized that it was a library that I had forked and included as my dependancy recently. I'm guessing that the prepare script isn't called when the package is included "naturally" through npm, but it is when it is included as a github repo. For the record, the package I forked was react-beautiful-dnd. Once I updated my fork to remove the prepare script, the phantom pipeline issues stopped appearing. THANK YOU!

**Comment by Yaxian** [comment]
try this way:
```
RUN mkdir .yarncache
RUN yarn install --cache-folder ./.yarncache
RUN rm -rf .yarncache
```

**Comment by esutton** [comment]
- Seeing this on my CI machine but not my dev machine with yarn 1.16.0
- I have seen problem with both yarn 1.16.0 and 1.17.0

#6312 suggested adding '--network-concurrency 1' which ** *seems* ** to work.
````
# https://github.com/yarnpkg/yarn/issues/6312
# Added '--network-concurrency 1'
# 
# Example Error: 
# error confusing-browser-globals-1.0.7.tgz: 
# Integrity check failed for "confusing-browser-globals" (computed integrity doesn't match our records

yarn install --network-concurrency 1
````

**Comment by TikiTDO** [comment]
@esutton Check my first comment in this thread [here](https://github.com/yarnpkg/yarn/issues/7212#issuecomment-493720324). You can probably get away without having to use the `network-concurrency` key if you find the offending lib.

**Comment by bitsal** [comment]
I'm very disappointed with this issue and had to switch to pure npm. I wonder why yarn community can't fix the issue so long.

**Comment by esutton** [comment]
- For now, I will stay with ````yarn install --network-concurrency 1````
- Uninstalling yarn and reverting to npm will be my next step.

@bitsal yarn has caused way too much time debugging build failures.  I do not recall having to do the same with npm which just seemed to work.

@TikiTDO Thanks for advice, however I am using react native which has too many packages and dependencies, to sort through which ones are problem packages.  While I enjoy a good puzzle, it is far less time consuming to revert to using npm and uninstall yarn.

**Comment by bitsal** [comment]
@esutton `yarn install --network-concurrency 1` doesn't fully help me
I still have errors like this:
```
yarn install v1.16.0

[1/4] Resolving packages...

[2/4] Fetching packages...

error An unexpected error occurred: "ENOENT: no such file or directory, open '.../.cache/yarn/v4/npm-spdx-exceptions-2.2.0-2ea450aee74f2a89bfb94519c07fcd6f41322977/node_modules/spdx-exceptions/.yarn-metadata.json'".

info If you think this is a bug, please open a bug report with the information provided in ".../yarn-error.log".

info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

script returned exit code 1
```

**Comment by FezVrasta** [comment]
I can confirm the issue seems to be related to a git dependency. But a workspace is not needed in my case to repro the problem.

https://github.com/yarnpkg/yarn/issues/7212#issuecomment-493720324

**Comment by cinderblock** [comment]
I think this is a dupe of #2629

**Comment by FezVrasta** [comment]
It looks different to me 🤔

**Comment by TikiTDO** [comment]
The `--ignore-scripts` suggestion may be relevant to the issue at hand, though it honestly looks like that issue is really a few problems rolled into one discussion

**Comment by dgobaud** [comment]
Think we're having this problem too details here https://github.com/yarnpkg/yarn/issues/2629#issuecomment-522871612 seems related to having a dependency hosted on github.com

**Comment by ohjames** [comment]
Hitting this constantly in our CI. No git dependencies whatsoever. We wanted to migrate from `npm` to `yarn` but this issue makes us doubt `yarn`'s viability.

**Comment by aalexgabi** [comment]
When you have more than 1 packages that install over Git (that have prepare scripts), this error happens almost every time and is not spurious anymore.

**Comment by aalexgabi** [comment]
Somewhere I found a workaround that seems to work: set --network-concurrency to 1
```
yarn install --network-concurrency 1
```

**Comment by insidewhy** [comment]
@aalexgabi Are you running multiple instances of `yarn` concurrently? Maybe you need something like:

```
yarn install --mutex network
```

**Comment by eskp** [comment]
Related https://github.com/yarnpkg/yarn/issues/2629

**Comment by dlio** [comment]
This issue appears to have resurfaced in yarn version 1.22.0  (MacOS).  The issue is persistent regardless of installation method (via brew, npm, or install script).  The offending package is a git repository with no scripts.

--network-concurrency and --ignore-scripts flags did not worked for me.

Downgrading yarn to 1.21.1 did resolve this issue, however.
`yarn policies set-version 1.21.1`

**Comment by buu700** [comment]
That fixed the problem for me, thanks @dlio! (`--network-concurrency 1` did not.)

Based on the errors I'm seeing in https://github.com/pierrec/node-lz4/issues/92, it seems that yarn is extracting the tgz files in an incorrect order, causing failures when they contain hard links?

**Comment by andrmoel** [comment]
Hi there.
I found a way how to reproduce this issue. The error is caused when you run a **parallel** `yarn install` at the same project. We faced this issue on our Jenkins server.

1. delete your node_modules folder
2. open 2 console tabs
3. ensure that you run `yarn install` **parallel** in both tabs

This will cause the following error.

```
error https://registry.yarnpkg.com/cssstyle/-/cssstyle-1.2.2.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, chmod '/home/amoeller/.cache/yarn/v6/npm-cssstyle-1.2.2-427ea4d585b18624f6fdbf9de7a2a1a3ba713077-integrity/node_modules/cssstyle/.eslintrc.js'"
```

Haven't found a solution yet.

**Comment by cinderblock** [comment]
@andrmoel If you're actually running two instances on yarn concurrently, there is are a couple ways to have yarn not interfere with itself. Take a look at the `--mutex` options for `yarn`.

However, I believe most people here are not intentionally running two concurrent instances of yarn and the `--mutex` options do not solve the problem.

**Comment by andrmoel** [comment]
@cinderblock that works great. Thanks a lot

**Comment by shark-horse** [comment]
Provided workarounds failed for me. This change to the offending dependency in package.json worked in my case:

```
-    "require-css": "guybedford/require-css#<hash>",
+    "require-css": "https://github.com/guybedford/require-css#<hash>",
```

Other dependencies also had the shorter github references, but I did not have to change those.

I arrived at that solution through obscure hints on Github, shotgun programming, and mysticism. No clue why it works. I am a newbie all over again and question my professionalism. HTH someone, or I'll question my sanity also.

**Comment by ohana54** [comment]
I was able to consistently reproduce the error (by deleting the yarn cache after every try).
This is my minimal setup:
main package - `A`
dependencies:
```
`A` -> `B`
    -> `C` -> `B`
```

The dependencies are listed with http urls, not versions - both for `A`'s package.json and `C`'s package.json.
Example:
```
{
	"name": "A",
	"dependencies": {
		"C": "http://remote-cache/C/1.0.0"
	}
}
```

Basically what happens, is that `B` is being downloaded twice to the same temporary cache folder. Simplified flow:
After downloading `B` for the first time, it is extracted to the temp cache folder.
*WHILE* the extraction is ongoing, yarn tries to download `B` again (because it is a dependency of `C`).
Before yarn starts to download it, it checks the temp cache folder to see if it's already there.
It does that by looking for a specific metadata file, that is written only after an extraction is complete.
Since the first `B` extraction is still ongoing, the metadata file is not there, and yarn assumes `B` is not in the temp cache.
The problem - yarn deletes the temp cache folder (I guess to handle a corrupt state) before it starts downloading `B` again.
Since the deletion happens during the extraction, we get the "Extracting tar content" error.

I was able to "fix" the problem by changing yarn's code to wait for ongoing download and extractions for packages. Meaning I keep the promise from the first `B` download process, and return it when we try to download it again, instead of having another download process interfering with the first one.
It's a hacky fix so I don't think submitting a PR with it will help, I'm mentioning it here in case someone is familiar with the codebase and provide a better idea.

**Comment by HitkoDev** [comment]
@ohana54 Can confirm, changing https://github.com/yarnpkg/yarn/blob/0e7133ca28618513503b4e1d9063f1c18ea318e5/src/fetchers/tarball-fetcher.js#L254-L255 to cache promises by `this.reference` solves the problem.

**Comment by HitkoDev** [comment]
Also the same problem occurs with local tarballs (and the same fix works), as described by @ohana54

**Comment by ohana54** [comment]
Thanks @HitkoDev for the validation.
This is a major blocker for us, as we started keeping tarballs from PR changes (without the need to publish them to the real registry), and having a link to these tarballs in package.json files is currently a no-go.
I can try to create a PR with a promise cache, although I'm still not sure it's good enough.

**Comment by HitkoDev** [comment]
That would be great, that way other devs get at least a rough idea what's going on and test different possible solutions.

**Comment by IUnknown68** [comment]
I just ran into the same issue, and the explanation https://github.com/yarnpkg/yarn/issues/7212#issuecomment-493720324 seems make sense. It fits our situation.

I was able to confirm this by "manually" caching the package(s) in question, so they are not fetched again:

1) When yarn errors, remove that package from the cache.
2) Let yarn fetch the package (by adding it to a dummy project).
3) Repeat from 1) for every package that gives the error.

Now your cache should be up2date, and no concurrent fetch/extracts should happen anymore (for those packages).

Of course that's not a solution, but until the error is fixed, it might help when you get stuck.

**Comment by mscrivo** [comment]
A workaround I've found that works for CI, without having to drop network concurrency to 1 and have slow builds, is to clone the repo of the problematic package, run yarn install it, then run yarn install on app that references that package. That seems to cache it and then it doesn't need to be fetched on the main yarn install and thus no race condition.

**Comment by dmitryguesty** [comment]
@ohana54 ,
Described by you flow/error possibly happens because of network concurrency race condition. 
This error happens to me running yarn install on windows with unstable internet connection.

Solved by:

adding **.yarnrc** file to the project with following line:

`network-concurrency 1`

**Comment by ohana54** [comment]
@dmitryguesty yes, you can see in my PR how the race condition is solved. Unfortunately for us it's not an option limiting the concurrency to 1.

**Comment by RoXuS** [comment]
We have this error to since we are on yarn (via lerna).
`lerna boostrap` produce this error randomly and only on our gitlab-ci:

````
error https://registry.yarnpkg.com/typescript/-/typescript-3.9.7.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, stat '/usr/local/share/.cache/yarn/v6/npm-typescript-3.9.7-98d600a5ebdc38f40cb277522f12dc800e9e25fa-integrity/node_modules/typescript/lib/lib.dom.d.ts'"
````

Any tips to detect which package produce this error? Any option works for us (concurrency, delete yarn.lock...).

**Comment by cinderblock** [comment]
@RoXuS the best stopgap solution I've found is to install the intermediate package, that also depends on TS, first. This can be to a completely different folder on the system as long as you use the same user. This will install that package in a cached area first and then subsequent installs used the cached version that doesn't cause this race condition.

**Comment by Techwolf12** [comment]
@ohana54 [Reply to](https://github.com/yarnpkg/yarn/issues/7212#issuecomment-637978197) 
Nice catch! This seems to be exactly the issue we run into. Empty cache and it tries to extract multiple times while downloading the new version.

**Comment by bdwain** [comment]
I was able to find the problematic package on my development computer (even though i don't experience the error on that computer). note my computer is a mac, so i can't say for certain this will work in windows or linux.

if you `cd $(yarn cache dir)/.tmp` you'll be taken to a folder where git modules are stored for local building. if you don't have many git modules stored there (i had 2) it's easy to look at all of them for a module with a `prepare` script in package.json. that is likely causing the issue (or one of multiple modules causing the issue).

if you want a more reliable way to find the exact module, i managed to edit my local installation of yarn to add a log statement.

```js
   console.log('foooooo', _this2.config.lockfileFolder); //i added this line
   _this2.reporter.step(curr, total, _this2.reporter.lang('resolvingPackages'), emoji.get('mag'));
```
adding the above line highlighted the exact module in the `.tmp` directory that was causing the issue

**Comment by mitchellwarr** [comment]
Have encountered this issue with a docker image on CI flow. The issue does not happen locally.

Fixed it by adding a .yarnrc file to the project with following line:
```network-concurrency 1```

Managed to track the issue down to a dependency with a prepare script of ```yarn build```

Its a private repo on our end. If this is expected to take a while to fix I'd love to know so that I can change my approach to that repo

**Comment by Jurpp** [comment]
> Have encountered this issue with a docker image on CI flow. The issue does not happen locally.
> 
> Fixed it by adding a .yarnrc file to the project with following line: `network-concurrency 1`
Thanks a lot, this is a valid (though very slow) workaround for the build issues we experienced when using AppCenter!

**Comment by U-4-E-A** [comment]
For me it was: -

"prepare": "husky install"

**Comment by Dan-Wood** [comment]
If anyone is experiencing this still, as we were on bitbucket.

We have two private repos that are imported into our main one, each of these repos have a prepare script that runs husky install.

The prepare script has been changed to the below
`"prepare": "node ./ci-skip-husky.js || husky install",`

And the script ci-skip-husky.js
`if (process.env.CI) {
	process.exit(0);
} else {
	process.exit(1);
}
`

A .yarnrc file with the current options set
`network-concurrency 1
child-concurrency 1
mutex network
`

**child-concurrency** being the main one.

This has fixed our issues on node 16.15.0, yarn 1.22.18

**Comment by jjeising** [comment]
> child-concurrency being the main one.

This option seems to be deprecated:

```
Unrecognized or legacy configuration settings found: child-concurrency
```

**Comment by Dan-Wood** [comment]
> > child-concurrency being the main one.
> 
> Note: these options were removed in yarn v2.

Original issue is using Yarn 1 not 2, still relevant.

**Comment by realtebo** [comment]
same problem appearing on a very fast fiber connection. why should be a problem of network concurrency?

**Comment by cinderblock** [comment]
This isn't caused by network issues. It's because multiple dependency trees overlap and Yarn doesn't do a great job of preventing a shared dependency with a build step from being built concurrently. It's a race condition that is apparently difficult to debug.

**Comment by markjm** [comment]
I don't think the debugging of the issue is the problem, its simply that yarn 1 is out of support. Seems there are 3 options:

1. upgrade to yarn 3
2. use this fork of yarn@1 which fixes the issue (and has other performance improvements)
3. Use your own fork of yarn, and apply this patch to it https://patch-diff.githubusercontent.com/raw/yarnpkg/yarn/pull/8182.patch

**Comment by realtebo** [comment]
We are using yarn 3, Same problem

**Comment by cengit** [comment]
in the pipeline `yarn install --network-concurrency 1` works very good for me
code snippet:
```
rm -rf ./yarn.lock
yarn install --network-concurrency 1
```

**Comment by ckxddd0324** [comment]
is there update for this?

**Comment by douglasjunior** [comment]
For us, `--network-concurrency 1` is enough on local machines, but is not in pipelines.

Like @Dan-Woo, we use many private packages installed directly from `git`.

Intermittent errors are:

```
error [https://registry.yarnpkg.com/es-abstract/-/es-abstract-1.20.5.tgz:](https://registry.yarnpkg.com/es-abstract/-/es-abstract-1.20.5.tgz) Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, stat '/usr/local/share/.cache/yarn/v6/npm-es-abstract-1.20.5-e6dc99177be37cacda5988e692c3fa8b218e95d2-integrity/node_modules/es-abstract/2015/SecFromTime.js'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

And

```
error An unexpected error occurred: "ENOENT: no such file or directory, open '/usr/local/share/.cache/yarn/v6/npm-omni-context-rn-0.0.1-a54c1dfa46bc463677451a88bf6ad83ad5ea6aaa/node_modules/my-private-library/.yarn-tarball.tgz'".
info If you think this is a bug, please open a bug report with the information provided in "/opt/atlassian/pipelines/agent/build/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by cinderblock** [comment]
I remember when Yarn was the new hotness. If you were still using Npm, everything took too long. Yarn was the only sane choice.

Then this happened. I dealt with it for a while with manual build scripts that manually installed dependencies from git sources in a temporary directory first - this seemed to reliably fix the problem but always felt kludgy. It was a bummer that Yarn has refused to fix the problem, or really even acknowledge that there is one, but at least I had a workaround and was happy enough to keep using Yarn.

Then Yarn 2 came crashing into the party with it's own headaches. Notably [intentionally breaking compatibility with many Npm lifecycle scripts](https://yarnpkg.com/advanced/lifecycle-scripts). As far as I can tell, there is now no way to have a "prepare" process that works consistently between Npm and Yarn - there is no way, that I'm aware of, to publish TypeScript sources such that when they're installed via git dependency, they consistently compile into the final desired JavaScript in both Yarn and Npm (without extra unreliable monkey patches in the build scripts).

We've had reliable repros here for **years**. How is this still a thing?

Is there a reason to keep using Yarn?

**Comment by arcanis** [comment]
> We've had reliable repros here for **years**. How is this still a thing?

You're two major versions behind, three if we count the current RCs for the next major.

We know the 1.x has a couple of rare but annoying bugs that are unfortunately quite hard to safely address. This is in large part why, almost four years ago now, we decided to publish new majors that tackle this problem. While using Yarn 1.x is still possible, it's nonetheless _advised_ to upgrade if you're hitting frustrating issues like this one.

As for problems with modern releases, I encourage you to open issues on their [dedicated tracker](https://github.com/yarnpkg/berry). I don't however remember hearing about problems similar to the ones you mention (the `prepack` script seems to do what you want, as described on its documentation page), so perhaps open a discussion first to be sure we're talking about the same thing.

In any case, I'll lock this thread as our team doesn't plan to merge bugfixes here. Understand that any change we make here has the potential to break many applications, so for the better or the worst we decided to simply freeze the old 1.x codebase (only exceptions being potential security fixes, and migration-related improvements).
